### PR TITLE
[TRACKER-258] - Add MetaData Change information to the Audit Log Screen

### PR DESCRIPTION
**Pull request by sushil-car**
https://issues.cask.co/browse/TRACKER-258

### CODE REVIEW

**Comment by ajainarayanan** [comment]
@sushiladokar Noticed two minor things,
- There is an extra text stroke on fa-times on the modal that we open
- If we don't have any metadata change do we need to have the link clickable to show an empty modal?

File: cdap-ui/app/tracker/controllers/audit-ctrl.js [hunk 1]
```diff
@@ -118,9 +119,120 @@ class TrackerAuditController {
...
+    schema= {},
+    properties = {};
+
+  let getSchemaArray = (schemaToParse) => {
+    if(!schemaToParse) {
+      return [];
+    }
+
+    let schemArr = [];
+    let parsed = avsc.parse(schemaToParse, { wrapUnions: true });
```
Comment (ajainarayanan) [inline cdap-ui/app/tracker/controllers/audit-ctrl.js]: @elfenheart can this throw an exception? Should we wrap this in a try catch block?

File: cdap-ui/app/tracker/controllers/audit-ctrl.js [hunk 2]
```diff
@@ -118,9 +119,120 @@ class TrackerAuditController {
...
+    let parsedSchema = parsed.getFields().map((field) => {
+      let type = field.getType();
+      let partialObj = SchemaHelper.parseType(type);
+      return Object.assign({}, partialObj, {
+        id: uuid.v4(),
+        name: field.getName()
+      });
+    });
+
+    parsedSchema.filter((obj) => {
```
Comment (ajainarayanan) [inline cdap-ui/app/tracker/controllers/audit-ctrl.js]: Should we just chain the `filter` to `.map`?

File: cdap-ui/app/tracker/controllers/audit-ctrl.js [hunk 3]
```diff
@@ -118,9 +119,120 @@ class TrackerAuditController {
...
+        schemArr.push(obj.name + ' ' + obj.displayType);
+      }
+    });
+
+    return schemArr;
+  };
+
+  let calculateUpdatedValues = (previous, additions, deletions) => {
+    let result = [];
+    result = _.union(previous, additions);
```
Comment (ajainarayanan) [inline cdap-ui/app/tracker/controllers/audit-ctrl.js]: 👍 Nice

File: cdap-ui/app/tracker/templates/audit.html [hunk 4]
```diff
@@ -58,6 +59,7 @@
             <span ng-if="log.type === 'ACCESS'">{{ log.payload.accessType }}</span>
           </td>
           <td>{{ log.user }}</td>
+          <td><a href="" ng-click="AuditController.showMetadataChangeInfo(log.payload);" ng-if="log.type === 'METADATA_CHANGE'">Details</a></td>
```
Comment (ajainarayanan) [inline cdap-ui/app/tracker/templates/audit.html]: nit: `<a />` tag in a newline

**Review by ajainarayanan — COMMENTED** [review]
Minor comments.

**Comment by ajainarayanan** [comment]
@russorat @sushiladokar Is this how we are showing previous and updated values? 
screen shot 2017-02-14 at 6 27 30 pm

What does `body string` mean? as in thats the schema it was updated with? Should we represent it in a schema editor if it can go complex?

File: cdap-ui/app/tracker/controllers/audit-ctrl.js (hunk 1 above)
Comment (elfenheart) [reply cdap-ui/app/tracker/controllers/audit-ctrl.js]: technically yes, but that would mean something broke in the backend

File: cdap-ui/app/tracker/controllers/audit-ctrl.js [hunk 5]
```diff
@@ -118,9 +119,120 @@ class TrackerAuditController {
...
+    result = _.difference(result, deletions);
+    return result;
+  };
+
+  if(logDataObj.previous.USER) {
```
Comment (elfenheart) [inline cdap-ui/app/tracker/controllers/audit-ctrl.js]: nit: space after `if`

Please go through this entire file

File: cdap-ui/app/tracker/controllers/audit-ctrl.js [hunk 6]
```diff
@@ -118,9 +119,120 @@ class TrackerAuditController {
...
+    tags.previousValues = logDataObj.previous.USER.tags;
+    if(logDataObj.additions.USER.tags.length > 0 || logDataObj.deletions.USER.tags.length > 0) {
+      tags.updatedValues = calculateUpdatedValues(logDataObj.previous.USER.tags, logDataObj.additions.USER.tags, logDataObj.deletions.USER.tags);
+    }
+
+    properties.previousValues = _.map(logDataObj.previous.USER.properties, (value, key) =>{
```
Comment (elfenheart) [inline cdap-ui/app/tracker/controllers/audit-ctrl.js]: space before `{`

File: cdap-ui/app/tracker/controllers/audit-ctrl.js [hunk 7]
```diff
@@ -118,9 +119,120 @@ class TrackerAuditController {
...
+
+    if(logDataObj.previous.SYSTEM.properties.schema) {
+      schema.previousValues = getSchemaArray(logDataObj.previous.SYSTEM.properties.schema);
+    }
+
+    let schemaArrayAddition = getSchemaArray(logDataObj.additions.SYSTEM.properties.schema);
+    let schemaArrayDeletion = getSchemaArray(logDataObj.deletions.SYSTEM.properties.schema);
+
+    schema.updatedValues = calculateUpdatedValues(schema.previousValues, schemaArrayAddition, schemaArrayDeletion);
+  }
```
Comment (elfenheart) [inline cdap-ui/app/tracker/controllers/audit-ctrl.js]: this block of code for `SYSTEM` seems to be very similar to `USER`. Can this be refactored into a helper function and consolidate the code for `SYSTEM` and `USER`?

File: cdap-ui/app/tracker/templates/partial/metadata-change-info.html [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  <table class="table table-bordered" ng-if="!MetadataInfo.noMetadataChange">
+    <thead>
+      <tr>
+        <th>Key</th>
+        <th>Previous Value</th>
+        <th>Updated Value</th>
+      </tr>
+    </thead>
+    <tbody>
+      <tr ng-repeat="(key, value) in MetadataInfo.formattedData" ng-if="value.previousValues.length >0 || value.updatedValues.length >0">
```
Comment (elfenheart) [inline cdap-ui/app/tracker/templates/partial/metadata-change-info.html]: nit: space after `>`

**Comment by elfenheart** [comment]
@russorat I left a comment on the JIRA: https://issues.cask.co/browse/TRACKER-258

This computation of the Updated Values should come from the app instead of being merged in the front end. It seems to be fundamentally wrong

File: cdap-ui/app/tracker/controllers/audit-ctrl.js [hunk 9]
```diff
@@ -118,9 +119,124 @@ class TrackerAuditController {
         });
       });
   }
+
+  isMetadataChanged(type, payload) {
+
+    if (type !== 'METADATA_CHANGE') {
+      return false;
+    } else {
```
Comment (elfenheart) [inline cdap-ui/app/tracker/controllers/audit-ctrl.js]: don't need to nest this under `else`

**Comment by elfenheart** [comment]
👍  LGTM

**Pull request merged**